File: tracker/lib/src/main/java/com/cyrillrx/tracker/AsyncTracker.java
package com.cyrillrx.tracker;

import com.cyrillrx.tracker.consumer.EventConsumer;
import com.cyrillrx.tracker.consumer.NamedThreadFactory;
import com.cyrillrx.tracker.event.TrackEvent;

import java.util.Queue;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

/**
 * @author Cyril Leroux
 *         Created on 19/04/16.
 */
public abstract class AsyncTracker<EventQueue extends Queue<TrackEvent>> extends TrackerChild {

    private static final String TAG = AsyncTracker.class.getSimpleName();

    private static final String THREAD_PREFIX = TAG + "_";
    protected static final int DEFAULT_WORKER_COUNT = 1;

    protected EventQueue pendingEvents;

    public AsyncTracker(EventQueue queue) { pendingEvents = queue; }

    /**
     * Adds the event to the queue where it will be consume.
     *
     * @param event The event to add to the queue.
     */
    @Override
    protected void doTrack(TrackEvent event) { pendingEvents.add(event); }

    /** Starts the pending events consumer(s). */
    protected void start() { start(DEFAULT_WORKER_COUNT); }

    /** Starts the pending events consumer(s). */
    protected void start(int workerCount) {

        final int processorCount = Runtime.getRuntime().availableProcessors();
        final ExecutorService executorService = Executors.newFixedThreadPool(
                processorCount,
                new NamedThreadFactory(THREAD_PREFIX));

        try {
            for (int i = 0; i < workerCount; ++i) {
                submitToService(executorService);
            }

        } finally {
            executorService.shutdown();
        }
    }

    protected void submitToService(ExecutorService service) { service.submit(createConsumer()); }

    protected abstract EventConsumer<EventQueue> createConsumer();
}

File: tracker/lib/src/androidMain/kotlin/com/cyrillrx/tracker/AsyncTracker.kt
package com.cyrillrx.tracker

import com.cyrillrx.tracker.consumer.EventConsumer
import com.cyrillrx.tracker.consumer.NamedThreadFactory
import com.cyrillrx.tracker.event.TrackEvent
import java.util.Queue
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors

/**
 * @author Cyril Leroux
 * Created on 19/04/16.
 */
abstract class AsyncTracker<EventQueue : Queue<TrackEvent>>(
    protected var pendingEvents: EventQueue
) : TrackerChild("AsyncTracker") {

    override fun doTrack(event: TrackEvent) {
        pendingEvents.add(event)
    }

    /** Starts the pending events consumer(s).  */
    protected fun start(workerCount: Int = DEFAULT_WORKER_COUNT) {
        val processorCount = Runtime.getRuntime().availableProcessors()
        val executorService = Executors.newFixedThreadPool(
            processorCount,
            NamedThreadFactory(THREAD_PREFIX)
        )
        try {
            for (i in 0 until workerCount) {
                submitToService(executorService)
            }
        } finally {
            executorService.shutdown()
        }
    }

    protected open fun submitToService(service: ExecutorService) {
        service.submit(createConsumer())
    }

    protected abstract fun createConsumer(): EventConsumer<EventQueue>?

    companion object {
        private val TAG: String = AsyncTracker::class.java.simpleName

        private val THREAD_PREFIX: String = TAG + "_"
        const val DEFAULT_WORKER_COUNT: Int = 1
    }
}
